lib/index.js

```diff
@@ -12,34 +26,312 @@ class PluginPassportOAuth {
 
-    if (Object.keys(config.strategies).length === 0) {
+    if (Object.keys(this.config.strategies).length === 0) {
+      /*eslint no-console: 0*/
       console.warn('kuzzle-plugin-auth-passport-oauth: no strategies specified.');
     }
 
-    this.config = config;
     this.context = context;
 
     Object.keys(this.config.strategies).forEach(key => {
       this.scope[key] = this.config.strategies[key].scope;
     });
 
-    if (!config.level) {
-      config.level = 'error';
+    return this.context.accessors.storage.bootstrap(storageMapping)
+      .then(() => {
+        return new Promise((resolve, reject) => {
+          Object.keys(this.config.strategies).forEach(key => {
+            if (!this.config.strategies[key].credentials) {
+              return reject(new this.context.errors.BadRequestError(`Error loading strategy [${key}]: no credentials provided`));
+            }
+
+            try {
+              this.strategyConstructor = require('passport-' + key).Strategy;
+            } catch (err) {
+              return reject(new this.context.errors.BadRequestError(`Error loading strategy [${key}]: ${err.message}`));
+            }
+          });
+          this.initStrategies();
+          return resolve();
+        });
+      });
+  }
+
+  afterRegister (strategyInstance) {
+    try {
+      refresh.use(strategyInstance);
+    } catch (err) {
+      /*eslint no-console: 0*/
+      console.error(`Error refreshing strategy: ${err.message}`);
     }
+  }
+
+  initStrategies () {
+    this.strategies = {};
+    Object.keys(this.config.strategies).forEach(strategy => {
+      this.strategies[strategy] = {
+        config: {
+          constructor: this.strategyConstructor,
+          strategyOptions: this.config.strategies[strategy].credentials,
+          authenticateOptions: {
+            scope: this.config.strategies[strategy].scope
+          },
+          fields: ['access_token', 'refresh_token']
+        },
+        methods: {
+          create: 'create',
+          delete: 'delete',
+          exists: 'exists',
+          getInfo: 'getInfo',
+          update: 'update',
+          validate: 'validate',
+          verify: 'verify',
+          afterRegister: 'afterRegister'
+        }
+      };
+    });
+  }
+
+  /**
+   * Get the repository according to the strategy
+   *
+   * @param provider
+   * @returns {module.exports.constructors.Repository|*}
+   */
+  getProviderRepository (provider) {
+    if (!this.providerRepository) {
+      this.providerRepository = new this.context.constructors.Repository(provider);
+    }
+
+    return this.providerRepository;
+  }
+
+  /**
+   * Get the credentials of a given kuid from the repository of the corresponding strategy
+   *
+   * @param kuid
+   * @param provider
+   * @returns {Promise|Promise.<TResult>}
+   */
+  getCredentialsFromKuid (kuid, provider) {
+    return this.getProviderRepository(provider).search({
+      query: {
+        match: {
```

Comment:
nitpicking: should be a [term](http://docs.kuzzle.io/elasticsearch-cookbook/#the-term-query) query instead of a [match](http://docs.kuzzle.io/elasticsearch-cookbook/#the-match-query) one.